kotest-assertions/kotest-assertions-core/src/jvmTest/kotlin/com/sksamuel/kotest/data/StringTableTest.kt

```diff
@@ -0,0 +1,213 @@
+package com.sksamuel.kotest.data
+
+import io.kotest.assertions.throwables.shouldThrowMessage
+import io.kotest.core.spec.style.FunSpec
+import io.kotest.data.Row3
+import io.kotest.data.headers
+import io.kotest.data.mapRows
+import io.kotest.data.row
+import io.kotest.data.table
+import io.kotest.data.toTable
+import io.kotest.data.writeTable
+import io.kotest.engine.spec.tempfile
+import io.kotest.matchers.shouldBe
+import java.io.File
+
+class StringTableTest : FunSpec({
+
+   val headers = headers("id", "username", "fullName")
+   val transform = { a: String, b: String, c: String ->
+      row(a.toInt(), b, c)
+   }
+   val expectedTable = table(
+      headers,
+      row(4, "jmfayard", "Jean-Michel Fayard"),
+      row(6, "louis", "Louis Caugnault"),
+   )
+
+   val validFileContent = """
+      4  | jmfayard | Jean-Michel Fayard
+      6  | louis    | Louis Caugnault
+   """.trimIndent()
+
+   context("creating tables from collections") {
+      test("create table from map") {
+         val map = mapOf(
+            "fr" to "French",
+            "es" to "Spanish"
+         )
+         val table = map.toTable(headers("code", "language"))
+         table shouldBe table(
+            headers("code", "language"),
+            row("fr", "French"),
+            row("es", "Spanish"),
+         )
+      }
+
+      val languagesTable = table(
+         headers("code", "name", "english"),
+         row("fr", "Français", "French"),
+         row("es", "Español", "Spanish"),
+      )
+
+      data class Language(val code: String, val english: String, val name: String)
+
+      val languages = listOf(
+         Language("fr", "French", "Français"),
+         Language("es", "Spanish", "Español"),
+      )
+
+      test("create table from list") {
+         val table = table(
+            headers("code", "name", "english"),
+            languages.map { row(it.code, it.name, it.english) }
+         )
+         table shouldBe languagesTable
+      }
+   }
+
+   test("happy path") {
+      table(headers, validFileContent, transform) shouldBe expectedTable
+   }
+
+   test("empty lines and comments starting with # are accepted") {
+      val fileContent = """
+         |4  | jmfayard | Jean-Michel Fayard
+         |# this is a comment
+         |# newlines are allowed
+         |
+         |6  | louis    | Louis Caugnault
+      """.trimMargin()
+      val table = table(headers, fileContent, transform)
+
+      table shouldBe expectedTable
+   }
+
+   test("All rows must have the right number of columns") {
+      val invalidRows = """
+      4  | jmfayard | Jean-Michel Fayard
+      5  | victor | Victor Hugo | victor.hugo@guernesey.co.k
+      6  | louis    | Louis Caugnault
+      7  | edgar
+   """.trimIndent()
+      shouldThrowMessage("Expected all rows to have size 3, but got rows at lines [1, 3]") {
```

Comment:
Perhaps we can build an even more elaborate failure message. Could help if/when there's rows with comments etc as well. Plus it removes ambiguity regarding 0 vs 1-indexing. Just an idea, feel free to skip if it adds too much scope.

```suggestion
      shouldThrowMessage("""
        Expected all rows to have 3 columns but some rows differed:
          - Row 1 has 4 columns ("5  | victor [...]")
          - Row 3 has 2 columns ("7  | edgar")
      """ ) {
```

Reply:
@Kantis my issue is more that there could be a large number of rows inside the file.
Should we take the first 5 with errors or something?

Reply:
@Kantis done